### ISSUE TITLE
Fonts with 1 typeface are not rendering on Android

### ISSUE THREAD

**Issue by oleksandr-dziuban** [eVisit/react-native-dynamic-fonts#15]
Hello @th317erd,
I have a problem with registering fonts correctly on Android. Could you please help? I noticed the next issue:

- If I register font with simple name, for example: "Archer" - Android rendering fonts correctly.

- If I register font with font style modificator in name, for example: OpenSansCondensed-Bold or Roboto-Italic or AngelineVintage-Medium - Android doesn't render custom form, text is looking default.

I always use postScriptName of font.
iOS app shows all fonts correctly.

Could you please check the ``ReactFontManager.getInstance().setTypeface(name, typeface.getStyle(), typeface);`` code in ``DynamicFontsModule.java``, I think the font is registering incorrectly if he has just one typeface, for example Bold or Italic or BoldItalic etc.

Example font which is not rendering by Android:
Name: "OpenSansCondensed-Bold",
Type: TTF
File: [font-file-with-base64.zip](https://github.com/eVisit/react-native-dynamic-fonts/files/3510300/font-file-with-base64.zip)
This ZIP archive contains TTF file and encoded base64 font. I use this base64 string dynamically and pass to `loadFont` as `data`.

Note: If I link this static font file with `react-native link` it works fine on iOS/Android, so problem in dynamic registering font in .java file.

I'm using:
react-native@0.60.5
react-native-dynamic-fonts@0.3.1

Best regards, Oleksandr

**Comment by th317erd** [comment]
@oleksandr-dziuban Have you tried using the "name" property for the font? i.e.

```
import { loadFonts } from 'react-native-dynamic-fonts';

await loadFonts([{
  "name": "OpenSansCondensed-Bold",
  "type": "ttf",
  "data": "AAEAAAATAQAABAAwRFNJR1p3/UAAAymkAAAVdEdERUYAJgOvAAMlhAAAAB5HUE9TCzcPNwADJaQAAAA4R1NVQg4rPbcAAyXcAAADxk9TLzKhRZY9AAABuAAAAGBjbWFwrrv1+wAAELQAAAOIY3Z0IA7AFxkAABz8AAAAoGZwZ21+YbYRAA..."
}])
```

On Android the "name" is an override of the internal font name.

**Comment by oleksandr-dziuban** [comment]
> @oleksandr-dziuban Have you tried using the "name" property for the font? i.e.
> 
> ```
> import { loadFonts } from 'react-native-dynamic-fonts';
> 
> await loadFonts([{
>   "name": "OpenSansCondensed-Bold",
>   "type": "ttf",
>   "data": "AAEAAAATAQAABAAwRFNJR1p3/UAAAymkAAAVdEdERUYAJgOvAAMlhAAAAB5HUE9TCzcPNwADJaQAAAA4R1NVQg4rPbcAAyXcAAADxk9TLzKhRZY9AAABuAAAAGBjbWFwrrv1+wAAELQAAAOIY3Z0IA7AFxkAABz8AAAAoGZwZ21+YbYRAA..."
> }])
> ```
> 
> On Android the "name" is an override of the internal font name.

@th317erd Hello, yes, I have this name "OpenSansCondensed-Bold" in my postScriptName in font and I pass it to name to load fonts too

**Comment by th317erd** [comment]
@oleksandr-dziuban I will look into this when I have a spare moment... which might be a little bit. The funny thing is that we are also using OpenSans-Bold, OpenSans-Italic, etc... with no issues. Maybe this is a React-Native 0.60 issue? At eVisit we haven't yet moved to 0.60

**Comment by oleksandr-dziuban** [comment]
@th317erd Nope, this issue is present on react-native <0.60 too.
It would be great if you check it when you have time, thanks a lot.

**Comment by oleksandr-dziuban** [comment]
Hello @@th317erd, do you have a few minutes to look into the issue with fonts on Android? It would be very great to fix this things... 
Thanks you a lot,
Best regards,
Oleksandr

**Comment by th317erd** [comment]
@oleksandr-dziuban No, I don't. I haven't forgotten about this (on my TODO list), but life is busy, and I likely won't be able to get to it for a few more weeks. If you want to look into fixing it and submit a PR I would be very grateful!

**Comment by oleksandr-dziuban** [comment]
@th317erd Sure, thank you
I don't have enough skills in Java to look into it. Thanks, will be waiting when you have some free time. Thanks

**Comment by oleksandr-dziuban** [comment]
@th317erd Hello, I have found the reason why some fonts are not rendering on Android. It was base64 encoding. I have changed way to fetch and generate base64 string with "rn-fetch-blob" and everything works fine now. I'll close this issue. Thanks

**Comment by th317erd** [comment]
OOhhhh, so your font wasn't base64 encoded? If that was what was causing the failure then I guess I need some better error handling at the least... :/

**Comment by oleksandr-dziuban** [comment]
> OOhhhh, so your font wasn't base64 encoded? If that was what was causing the failure then I guess I need some better error handling at the least... :/

No, it was encoded, but I encoded font with "react-native-fs" readFile(..., 'base64') method.
I think it doesn't encode correctly some fonts.
With "rn-fetch-blob" everything is fine. 
Just i'm fetching remote fonts dynamically from file server, and need to decode them on-a-flight.
Your error handling is fine, no changes are needed.

**Comment by oleksandr-dziuban** [comment]
Hello @th317erd,
Sorry, but issue is still reproducible
I will reopen this ticket.

Could you please take a look on it when you have time?

Best Regards,
Oleksandr

**Comment by th317erd** [comment]
Sure. I might have a chance to look at it in the next few days. I will keep you posted.

**Comment by oleksandr-dziuban** [comment]
@th317erd Thank you

**Comment by th317erd** [comment]
@oleksandr-dziuban Okay, I have been debugging for a while and I think I know what is wrong. The bad news is that I don't have a good programmatic fix. I also could not get your font working for quite some time until I dug deep into React Native code. As far as I can tell, React Native combines font family with font weight and style. Font weight is calculated as either "bold" or "regular" no matter what (these are the ONLY two internal weights). If you specify `fontStyle: 'italic'` it will also apply the "italic" modifier. What does this mean? This means that it will only match on your font if the "font family", "font weight", and "font style" all match. I finally got your font working by specifying `fontSize: 'bold'` (fontSize: '700' is equivalent inside the code).

So, this means that you can not have the full range of weights inside a font family. In the app I work on we have OpenSans-Regular, OpenSans-SemiBold, and OpenSans-Bold. OpenSans-SemiBold doesn't work if we were to load all the fonts named as the font family "OpenSans", because again, react native only has four states: normal, bold, italic, bold_italic. This means if I register OpenSans with ALL the weights it provides, the only option I have is to register **AND** use them with the full font name **INCLUDING** the weight. For example, you would **NOT** do this (pseudo code just to explain):
`loadFonts([ OpenSansRegular as OpenSans, OpenSansSemiBold as OpenSans, OpenSansBold as OpenSans ]);` `{ fontFamily: 'OpenSans', fontWeight: '400' }` (intending SemiBold variant) because, even though the fonts would technically register, React Native would not be able to find them or use them this way (because SemiBold doesn't match any of the four internal font-weights).

In our situation we are instead registering them with the full name, and using them with the full name (again, pseudo code just to show what I mean):
`loadFonts([ OpenSansRegular as OpenSansRegular, OpenSansSemiBold as OpenSansSemiBold, OpenSansBold as OpenSansBold ]);` `{ fontFamily: 'OpenSans-SemiBold', fontWeight: 'normal' }`, `{ fontFamily: 'OpenSans-Bold', fontWeight: 'bold' }`. We make this easier on ourselves in our dev team by making font constants we can then spread into a style: `FONT_SEMIBOLD = { fontFamily: 'OpenSans-SemiBold', fontWeight: 'normal' }` and later `someTextStyle: { ...FONT_SEMIBOLD }`

This now works, because we ignore the weight (more or less), and register each typeface weight under its own name. Notice though how I still specify the MATCHING font-weight for the typeface in question. If you don't, you again risk it not matching the typeface you requested (i.e. every weight under "bold" should be "normal", and every weight above "bold" should be "bold").

I might be wrong in some of the information I just specified, but this is how I understand it from reading the React Native code, and from testing different scenarios.

TL;DR; I was able to get your typeface to work by specifying `fontWeight: 'bold'` along with the named `fontFamily: 'OpenSansCondensed-Bold'`.

**Comment by oleksandr-dziuban** [comment]
@th317erd Thanks a lot for your deep research!
So, we just need to specify additionally ```fontWeight: bold```, or ```fontStyle: italic``` for Bold/Italic specific fonts right?

**Comment by th317erd** [comment]
@oleksandr-dziuban Yes, for bold and italic fonts I believe you will need to specify the 'bold' `fontWeight` and 'italic' `fontStyle` or you risk them not showing up.

It would be awesome if you could open an issue on the react-native project and ask this question. It appears to me from reading the code that there is better font support IF the fonts are bundled as assets with the application, but not when they are dynamically loaded like this. In any case, getting some feedback from someone who actually knows more about the font system would indeed be helpful.

After we get said feedback, and I have a better understanding of what is happening (and am confident in that understanding) I can update the documentation for this project to reflect the new information.

**Comment by oleksandr-dziuban** [comment]
@th317erd Thank you for your research, you are right, for Android we need additionally assign fontWeight, fontStyle and fontStretch props to correctly render font.
I have implemented font family parsed with maps for all font variants with related font styles and everything looks good now. Thanks again! Closing issue..

**Comment by th317erd** [comment]
Awesome! Good to hear! Let me know if there is any other way I can help out :)

**Comment by th317erd** [comment]
Great article that explains a lot of nuances: https://medium.com/@lewie9021/custom-fonts-in-react-native-85d814ca084

**Comment by oleksandr-dziuban** [comment]
@th317erd Thanks, will check it 👍
